### PR TITLE
feat(AIP-121): enforce list requirement

### LINKED ISSUE #1104: AIP-121: require GET and LIST
AIP-121 states: 

> A resource must support at minimum [Get](https://google.aip.dev/131): clients must be able to validate the state of resources after performing a mutation such as [Create](https://google.aip.dev/133), [Update](https://google.aip.dev/134), or [Delete](https://google.aip.dev/135).

> A resource must also support [List](https://google.aip.dev/132), except for [singleton resources](https://google.aip.dev/156) where more than one resource is not possible.

### PR DESCRIPTION
**Pull request by noahdietz**
Fixes #1104

### CODE REVIEW

File: docs/rules/0121/resource-must-support-get.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 rule:
   aip: 121
   name: [core, '0121', resource-must-support-get]
-  summary: All resource names must use camel case in collection identifiers.
+  summary: All resources names must have a Standard Get.
```
Comment (toumorokoshi) [inline docs/rules/0121/resource-must-support-get.md:5]:
```suggestion
  summary: All resources must have a Standard Get method.
```

File: docs/rules/0121/resource-must-support-list.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+---
+rule:
+  aip: 121
+  name: [core, '0121', resource-must-support-list]
+  summary: All resources names must have a Standard List.
```
Comment (toumorokoshi) [inline docs/rules/0121/resource-must-support-list.md:5]:
```suggestion
  summary: All resources must have a Standard List method.
```

File: rules/aip0121/resource_must_support_list.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+					resourcesWithList.Add(t)
+				}
+			} else if utils.IsCreateMethod(m) || utils.IsUpdateMethod(m) || utils.IsGetMethod(m) {
+				if msg := utils.GetResponseType(m); msg != nil {
+					// Skip tracking Singleton resources, they do not need List.
+					if utils.IsSingletonResource(msg) {
+						continue
+					}
+					t := utils.GetResource(msg).GetType()
+					resourcesWithOtherMethods[t] = msg
```
Comment (toumorokoshi) [inline rules/aip0121/resource_must_support_list.go:49]: nit: any reason not to do a stringset here? it looks like the MessageDescriptor value is unused.

File: rules/aip0121/resource_must_support_list_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+			rpc UpdateBook(UpdateBookRequest) returns (Book) {};
+		`, []lint.Problem{
+			{Message: `resource "library.googleapis.com/Book"`},
+		}},
+		{"InvalidGetOnly", `
+			rpc GetBook(GetBookRequest) returns (Book) {};
+		`, []lint.Problem{
+			{Message: `resource "library.googleapis.com/Book"`},
+		}},
+		{"IgnoreSingleton", `
```
Comment (toumorokoshi) [inline rules/aip0121/resource_must_support_list_test.go:78]:
```suggestion
		{"ValidIgnoreSingleton", `
```

nit really, but most other tests only categorize under `Valid` and `Invalid`, not `Ignore`.

**Review by toumorokoshi — APPROVED** [review]
Thanks! LGTM with a few nits.

File: rules/aip0121/resource_must_support_list.go (hunk 3 above)
Comment (noahdietz) [reply rules/aip0121/resource_must_support_list.go:49]: Mm I copied the code from the `resource_must_support_get` rule and just adjusted it. I can change this though.

File: rules/aip0121/resource_must_support_get.go [hunk 5]
```diff
@@ -26,11 +26,10 @@ import (
 var resourceMustSupportGet = &lint.ServiceRule{
 	Name: lint.NewRuleName(121, "resource-must-support-get"),
 	LintService: func(s *desc.ServiceDescriptor) []lint.Problem {
-		problems := []lint.Problem{}
-		// Add the empty string to avoid adding multiple nil checks.
-		// Just say it's valid instead.
-		resourcesWithGet := stringset.New("")
```
Comment (noahdietz) [inline rules/aip0121/resource_must_support_get.go:32]: FYI @toumorokoshi we don't need to init with empty string, we can init as `nil` and the `stringset.Set` methods will handle when the receiver is `nil`/empty https://go.dev/play/p/2BXM0MTu1i7

**Pull request merged**